File: frontend/src/types/analysis.ts
/**
 * Part 1: 用户画像分析 — 全局类型定义
 * 严格对齐 docs/02-part1-frontend.md 的 API 数据结构
 */

// ========== 任务相关 ==========
export type TaskStatus = 'pending' | 'collecting' | 'analyzing' | 'waiting_tags' | 'processing' | 'completed' | 'failed';

export interface AnalysisTask {
  taskId: string;
  bloggerName: string;
  bloggerAvatar?: string;
  platform: string;
  analysisLevel: '快速' | '标准' | '深度' | '高级自定义';
  status: TaskStatus;
  progress: number;            // 0-100
  currentStep?: string;
  createdAt: string;
  completedAt?: string;
  errorMessage?: string;
}

// ========== 画像报告相关 ==========

/** 单个子维度的图表数据项 */
export interface ChartDataItem {
  name: string;
  value: number;     // 百分比 (0-100) 或计数
  color?: string;    // 可选的自定义颜色
}

/** 子维度 */
export interface SubDimension {
  subId: string;
  subName: string;
  chartType: 'pie' | 'bar' | 'radar' | 'text';
  data: ChartDataItem[];
  summary?: string;           // 子维度的文字小结
}

/** 主维度 */
export interface Dimension {
  dimensionId: string;
  dimensionName: string;
  icon: string;               // emoji 或 icon key
  subDimensions: SubDimension[];
  overallSummary: string;     // 维度总结文字
}

/** 完整画像报告 */
export interface ProfileReport {
  taskId: string;
  bloggerName: string;
  bloggerAvatar?: string;
  platform: string;
  followerCount: string;      // "12.3万"
  analysisLevel: string;
  analysisTime: string;       // "2026-06-13 10:00"
  overallSummary: string;     // 综合画像总结
  dimensions: Dimension[];
  fullReportMarkdown: string; // 完整的 Markdown 文字报告
}

// ========== 标签筛选相关 ==========

export interface Tag {
  name: string;
  percentage: number;         // 0-100
  isDefaultSelected: boolean;
}

export interface TagSubDimension {
  subId: string;
  subName: string;
  tags: Tag[];
  isMutuallyExclusive: boolean;
  maxSelect: number | null;   // null = 无限制
}

export interface TagDimension {
  dimensionId: string;
  dimensionName: string;
  icon: string;
  subDimensions: TagSubDimension[];
}

// ========== 创建分析请求 ==========
export interface CreateAnalysisRequest {
  bloggerUrl: string;
  platform: string;
  analysisLevel: string;
  customConfig?: {
    hotPostCount?: number;
    recentPostCount?: number;
    sortBy?: string;
    commentsPerPost?: number;
    commentSort?: string;
    enableSubComments?: boolean;
    subCommentsPerComment?: number;
    maxCommenterAnalysis?: number;
    analyzeCommenterPosts?: boolean;
    commenterPostCount?: number;
    analyzeCoverImages?: boolean;
    analyzeVideoFrames?: boolean;
    framesPerVideo?: number;
    analyzedFramesPerVideo?: number;
    fanCoverAnalysis?: string;
    collageSize?: number;
  };
}

File: backend/app/schemas/analysis.py
"""
Analysis API Pydantic schemas (Part1 §2 接口契约).

R2 Three-Question Self-Check:
1. Contract Closure: Each request/response schema maps 1:1 to a documented API
   contract in docs/01-part1-backend.md §2.1-2.7. URL/level validation raises
   clean ValueError that the API layer translates to 400/422.
2. Symmetry: Pure DTO models, no resources to acquire/release.
3. External Timing: Stateless validation; safe for concurrent requests.
"""

import datetime
from typing import Any, Optional
from pydantic import BaseModel, Field, field_validator


SUPPORTED_PLATFORMS = {"douyin", "xiaohongshu", "taobao"}
PRESET_LEVELS = {"quick", "standard", "deep"}


class AnalysisCreateRequest(BaseModel):
    """POST /api/v1/analysis/create 请求体（§2.1）。"""

    blogger_url: str = Field(..., description="博主主页 URL，必须以 http 开头")
    platform: str = Field("auto", description="平台：douyin/xiaohongshu/taobao/auto")
    analysis_level: str = Field("standard", description="quick/standard/deep/custom")
    custom_config: Optional[dict[str, Any]] = Field(
        None, description="自定义配置，仅 analysis_level=custom 时必填"
    )

    @field_validator("blogger_url")
    @classmethod
    def _validate_url(cls, v: str) -> str:
        if not v or not isinstance(v, str) or not v.startswith("http"):
            raise ValueError("Invalid blogger URL format")
        return v


class AnalysisCreateData(BaseModel):
    task_id: str
    status: str = "pending"
    created_at: datetime.datetime


class BloggerInfo(BaseModel):
    nickname: Optional[str] = None
    avatar_url: Optional[str] = None
    follower_count: Optional[int] = None
    platform: Optional[str] = None


class AnalysisTaskDetail(BaseModel):
    """GET /api/v1/analysis/{task_id} 响应 data（§2.2）。"""

    task_id: str
    platform: str
    blogger_url: str
    analysis_level: str
    status: str
    progress: int
    current_step: str
    blogger_info: Optional[BloggerInfo] = None
    created_at: datetime.datetime
    updated_at: datetime.datetime
    completed_at: Optional[datetime.datetime] = None
    custom_config: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None


class AnalysisListItem(BaseModel):
    task_id: str
    platform: str
    blogger_info: Optional[BloggerInfo] = None
    analysis_level: str
    status: str
    progress: int
    created_at: datetime.datetime
    completed_at: Optional[datetime.datetime] = None
    has_fragrance_session: bool = False


class AnalysisListData(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[AnalysisListItem]


class AnalysisReportData(BaseModel):
    """GET /api/v1/analysis/{task_id}/report 响应 data（§2.4）。"""

    task_id: str
    blogger_info: BloggerInfo
    report: dict[str, Any]
    full_report_markdown: str


class TagItem(BaseModel):
    name: str
    percentage: float
    is_default_selected: bool = False
    mutually_exclusive_group: Optional[str] = None


class SubDimension(BaseModel):
    sub_id: str
    sub_name: str
    tags: list[TagItem]
    is_mutually_exclusive: bool = False
    max_select: Optional[int] = None


class Dimension(BaseModel):
    dimension_id: str
    dimension_name: str
    sub_dimensions: list[SubDimension]


class AnalysisTagsData(BaseModel):
    """GET /api/v1/analysis/{task_id}/tags 响应 data（§2.5）。"""

    dimensions: list[Dimension]


class CancelTaskData(BaseModel):
    task_id: str
    status: str = "cancelled"


class DeleteTaskData(BaseModel):
    task_id: str
    deleted: bool = True
